Added main menu screen for single/multiplayer selection

brendan-brooks opened the pull request:
- Added `fillRectBorder()` in `gfx.c` to draw thick borders for buttons
- Added `gui_draw_multiplayer_button()` and `gui_draw_singleplayer_button()` in `battleship_utils.c`
- Added `gui_draw_main_menu()` in `battleship_utils.c` to render the main menu screen
- Introduced new game states: `GS_RESET` (`handle_reset()`), `GS_MAINMENU` (`handle_main_menu()`), `GS_NEWGAME` (`handle_new_game()`)
- Moved `protocol_reset()` logic into `handle_reset()`; game now transitions to `GS_MAINMENU` after a reset call

pwkamp commented on AVRmada/src/main.c:
```diff
@@ -87,6 +87,9 @@ static inline void tx_result(uint8_t r, uint8_t c, bool hit) {
  * ------------------------------------------------------------------------- */
 static enum {
+	GS_RESET,
+	GS_MAINMENU,
```
Indentation does not match the rest of the enum

pwkamp commented on AVRmada/src/main.c:
```diff
@@ -541,10 +585,9 @@ int main(void) {
-    protocol_reset();        // Reset game state and draw placement UI
-
+	
+	gState = GS_RESET;		 // The initial game state is GS_RESET
```
indent

pwkamp commented on AVRmada/include/gfx.h:
```diff
@@ -125,6 +125,7 @@ uint16_t rgb(uint8_t r, uint8_t g, uint8_t b);
 // Shape drawing functions
 void drawRect(int16_t x, int16_t y, int16_t w, int16_t h, uint16_t color);
 void fillRect(int16_t x, int16_t y, int16_t w, int16_t h, uint16_t color);
```
Overload drawRect instead of calling a new function. I would make:

`void drawRect(int16_t x, int16_t y, int16_t w, int16_t h, uint16_t color);` be an alias to:
`drawRect(int16_t rect_x, int16_t rect_y, int16_t rect_w, int16_t rect_h, int16_t border_size, uint16_t color);`
with a default value for `border_size`

pwkamp commented on AVRmada/src/battleship_utils.c:
```diff
@@ -248,20 +248,64 @@ void enemy_place_random(void) {
+	gui_draw_singleplayer_button(CLR_LIGHT_GRAY, CLR_DARK_GRAY);
+	
+	// Finish the title ("AVRmada" found its letter V!)
+	for (uint8_t i=0; i<255; i+=3) {
```
Make this into a function

pwkamp commented on AVRmada/src/main.c:
```diff
@@ -105,26 +108,22 @@ static enum {
     NS_GAME_OVER
 } nState;
 
+static enum {
```
Move above gState since you are concerned with this before anything else

pwkamp commented on AVRmada/src/main.c:
```diff
@@ -105,26 +108,22 @@ static enum {
+
 /* -------------------------------------------------------------------------
- *  GAME INITIALIZATION / RESET
+ *  UI HANDLERS
```
Is this accurate? Only UI handling or does something else happen in these functions as well?

pwkamp commented on AVRmada/src/main.c:
```diff
@@ -553,6 +596,15 @@ int main(void) {
+				break;
+			case GS_NEWGAME:
+				handle_new_game();			// Draw initial game screen; goes to GS_PLACING
+				break;
```
Indentation does not match rest of switch statement

pwkamp commented on AVRmada/src/gfx.c:
```diff
@@ -556,6 +556,27 @@ void fillRect(int16_t x, int16_t y, int16_t w, int16_t h, uint16_t color) {
+ * Fill *just* the border of a rectangle using 4 strokes, each of thickness `border_size`
+ * (the *center* of the drawn strokes are aligned with the original rectangle)
+ */
+void fillRectBorder(int16_t rect_x, int16_t rect_y, int16_t rect_w, int16_t rect_h, int16_t border_size, uint16_t color) {
```
See header comment

pwkamp reviewed (changes requested):
Looks really good functionally! Logic is sound and passes manual tests.

I have left comments on changes that I would like made for clarity.

pwkamp reviewed (approved):
Good enough!

brendan-brooks commented:
add main menu screen

  - added `fillRectBorder()` in `gfx.c` to draw thick borders for buttons
  - added `gui_draw_multiplayer_button()` and `gui_draw_singleplayer_button()` in `battleship_utils.c`
  - added `gui_draw_main_menu()` in `battleship_utils.c` to render the main menu screen
  - Introduced new game states: `GS_RESET` (`handle_reset()`), `GS_MAINMENU` (`handle_main_menu()`), `GS_NEWGAME` (`handle_new_game()`)
  - moved `protocol_reset()` logic into `handle_reset()`; game now transitions to `GS_MAINMENU` after a reset call
  - added header definition `CLR_MM_BG` for the main menu's background color, in `battleship_utils.h`

brendan-brooks commented:
add main menu screen

  - added `fillRectBorder()` in `gfx.c` to draw thick borders for buttons
  - added `gui_draw_multiplayer_button()` and `gui_draw_singleplayer_button()` in `battleship_utils.c`
  - added `gui_draw_main_menu()` in `battleship_utils.c` to render the main menu screen
  - Introduced new game states: `GS_RESET` (`handle_reset()`), `GS_MAINMENU` (`handle_main_menu()`), `GS_NEWGAME` (`handle_new_game()`)
  - moved `protocol_reset()` logic into `handle_reset()`; game now transitions to `GS_MAINMENU` after a reset call
  - added header definition `CLR_MM_BG` in `battleship_utils.h` for the main menu's background color
